### <p style="text-align: right;"> &#9989; Ishan Baweja</p>
### <p style="text-align: right;"> &#9989; Ishanb314</p>

# CMSE 202 Midterm

The goal of this midterm is to give you the opportunity to test out some of the skills that you've developed thus far this semester. In particular, you'll practice setting up a GitHub repository, committing and pushing repository changes, downloading data with command line tools, performing some data analysis, possibly using a new Python package, and writing a python class. You should find that you have all of the skills necessary to complete this exam with even just eight weeks of CMSE 202 under your belt!

You are encouraged to look through the entire exam before you get started so that you can appropriately budget your time and understand the broad goals of the exam. Once you've read through it, try doing Parts 1 and 2 first so that you have your repository set up and you download all necessary data files as they will be necessary to complete the assigned tasks. Let your instructor know right away if you have problems downloading the data!

The exam is set up so that even if you get stuck on one part there are opportunities to get points on the other parts, so consider jumping ahead if you feel like you aren't making progress and then come back later if you have time.

**Important note about using online resources**: This exam is "open internet". That means that you can look up documentation, google how to accomplish certain Python tasks, etc. Being able to effectively use the internet for computational modeling and data science is a very important skill, so we want to make sure you have the opportunity to exercise that skill. **However**: The use of any person-to-person communication software is absolutely not acceptable. If you are seen accessing your email, using a chat program (e.g. Slack), or any sort of collaborative cloud storage or document software (e.g. Google Documents), you will be at risk for receiving a zero on the exam.

**Keep your eyes on your screen!** Unfortunately, there isn't enough space in the room for everyone to sit at their own table so please do your best to keep your eyes on your own screen. This exam is designed to give *you* the opportunity to show the instructor what you can do and you should hold yourself accountable for maintaining a high level of academic integrity. If any of the instructors observe suspicious behavior, you will, again, risk receiving a zero.

The maximum number of points is: **XX Points**

![Moira](https://media.giphy.com/media/26gs78HRO8sOuhTkQ/giphy.gif)

---
# &#128721; READ EVERYTHING CAREFULLY

---
# Part 0: Academic integrity statement (1 point)

Read the following statement and edit the markdown text to put your name in the statement. This is your commitment to doing your own authentic work on this exam.

> I, **INSERT NAME HERE**, affirm that this exam represents my own authetic work, without the use of any unpermitted aids or resources or person-to-person communication. I understand that this exam an an opportunity to showcase my own progress in developing and improving my computational skills and have done my best to demonstrate those skills.

---
# Part 1: Git and Command Line Interface (9 points)

Before you get to far along in the exam, you're going to add it to the `cmse202-XX-turnin` repository you created in class so that you can track your progress on the exam and preserve the final version that you turn in. 


---
**&#9989; **Question 1.1 (1 point)**** Navigate to your `cmse202-XX-turnin` repository and create a new directory called `midterm`. 

```
mkdir midterm
```

---
&#9989; **Question 1.2 (3 points)** Check the status of `git` .

``` 

```

Copy and paste below the output of the command.

```
Paste it here
```

What is the name of the branch you are in ? 

```
Put your answer here
```

**Important:** If you are not in the `main` branch change to your `main` branch. 

Once in the main branch move this notebook into that **new directory** in your repository.

---
&#9989; **Question 1.3 (3 points):**
Add your name and GitHub username to the top of the notebook, then add and commit **ONLY** the notebook.

```
Put the command(s) to add and commit here 
```

What is the commit message you used ?

```
Answer the question here
```

---
&#9989; **Question 1.4 (1 point):** Before moving on. Check that the notebook you are working on is the correct one. Run the following cell. **Are you in the new folder you just created?** If not close this notebook and open the one in the midterm folder


In [1]:
!pwd

/home/bawejais/cmse202-XX-turnin/midterm


---
&#9989; **Question 1.5 (1 point):**
Finally push the updated repo to GitHub.

```
Put the command you used to push to GitHub here.
```


---
# Part 2: Downloading and analyzing unfamiliar data (21 points)

For this part of the exam, you're to download and analyze a dataset that you've likely not looked at before. You'll perform some simple, exploratorary analysis and create basic visualizations.

In particular, you're going to be working with a dataset that contains data on all nuclear explosions between 1945 and 1998. The data are from `N.-O. Bergkvist and R. Ferm, Nuclear Explosions 1945–1998, Swedish Defence Research Establishment/SIPRI, Stockholm, July 2000`. 

Download the file from 

`https://raw.githubusercontent.com/cmse202-S22-data/main/code_samples/nuclear-explosion-data.csv`


&#9989; **Question 2.1 (1 point)**: **Do this now:** Save this files in the same directory as your notebook so you can load it directly. 

**Then, in the cell below, put the command or commands you used to download the files. If you did not use a command line tool to download the files, write down a command that *would* have fetched the files.**

``` 
Put the command(s) you used for downloading the data files here!

```

&#9989; **Question 2.2 (2 points)**: To get started, **read in the `nuclear-explosion-data.csv` dataset** and then display the first 6 rows of the data using **Pandas**.

In [15]:
# Put your code here
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

all_data_raw = pd.read_csv("nuclear-explosion-data.new.csv",lineterminator='\n')
all_data_raw.head(6)

date      time     id country      region source    lat    long   mb  \
0  19450716  123000.0  45001     USA  ALAMOGORDO    DOE  32.54 -105.57  0.0   
1  19450805  231500.0  45002     USA   HIROSHIMA    DOE  34.23  132.27  0.0   
2  19450809   15800.0  45003     USA    NAGASAKI    DOE  32.45  129.52  0.0   
3  19460630  220100.0  46001     USA      BIKINI    DOE  11.35  165.20  0.0   
4  19460724  213500.0  46002     USA      BIKINI    DOE  11.35  165.20  0.0   
5  19480414  181700.0  48001     USA    ENEWETAK    DOE  11.30  162.15  0.0   

    Ms  depth  yield_lower  yield_upper purpose       name     type\r  
0  0.0  -0.10         21.0         21.0      WR    TRINITY    TOWER\r  
1  0.0  -0.60         15.0         15.0  COMBAT  LITTLEBOY  AIRDROP\r  
2  0.0  -0.60         21.0         21.0  COMBAT     FATMAN  AIRDROP\r  
3  0.0  -0.20         21.0         21.0      WE       ABLE  AIRDROP\r  
4  0.0   0.03         21.0         21.0      WE      BAKER       UW\r  
5  0.0  -0.08         37.0         37.0      WR      X-RAY    TOWER\r

This dataset contains more information that we care about at the moment. 

&#9989; **Question 2.2 (4 pts)**: Drop all the columns **except** `date, country, lat, long, yield_upper, purpose, type`


In [25]:
# Put your code here
df = all_data_raw.drop(['time','id','region','source','mb','Ms','depth','yield_lower','name'], axis = 1)
df.head(6)

date country    lat    long  yield_upper purpose     type\r
0  19450716     USA  32.54 -105.57         21.0      WR    TOWER\r
1  19450805     USA  34.23  132.27         15.0  COMBAT  AIRDROP\r
2  19450809     USA  32.45  129.52         21.0  COMBAT  AIRDROP\r
3  19460630     USA  11.35  165.20         21.0      WE  AIRDROP\r
4  19460724     USA  11.35  165.20         21.0      WE       UW\r
5  19480414     USA  11.30  162.15         37.0      WR    TOWER\r

In [29]:
# Run this cell
# Change the df to the name of your dataframe. 
df["date"] = pd.to_datetime(df["date"], format = "%Y%m%d", errors='coerce')
df.head(5)

date country    lat    long  yield_upper purpose     type\r
0 1945-07-16     USA  32.54 -105.57         21.0      WR    TOWER\r
1 1945-08-05     USA  34.23  132.27         15.0  COMBAT  AIRDROP\r
2 1945-08-09     USA  32.45  129.52         21.0  COMBAT  AIRDROP\r
3 1946-06-30     USA  11.35  165.20         21.0      WE  AIRDROP\r
4 1946-07-24     USA  11.35  165.20         21.0      WE       UW\r

&#9989; **Question 2.2 (2 points)**: Use built-in Pandas functions to get a quick statistical summary of your data (e.g. the mean, standard deviation, min, max, etc) **divided by country .**

In [30]:
# Put your code here
df.describe()

lat         long   yield_upper
count  2051.000000  2051.000000   2046.000000
mean     35.397331   -36.048777    323.431021
std      23.403654   100.864340   2055.203066
min     -49.500000  -169.320000      0.000000
25%      37.000000  -116.054000     18.250000
50%      37.100000  -116.000000     20.000000
75%      49.870000    78.000000    150.000000
max      75.100000   179.220000  50000.000000

&#9989; **Question 2.2.1 (2 points)**: Looking at the above output, answer the following questions: 

1. What countries have used nuclear weapons?
2. How many weapons did the USA use?

**<font size=+3>&#9998;</font> Do This**: *Record your answers here.*



&#9989; **Question 2.3 (7 points)**: Using `seaborn` make an histogram ([_Hint_](https://seaborn.pydata.org/examples/index.html) ) : 

1. with the `date` as the `x` axis

2. set the `bins` option equal to 45

3. set the `hue` option equal to `country`. 

4. make the figure size equal to 12, 10

5. set the `palette` option equal to `"Dark2"`. Or changing it to your preference [Link to colors](https://matplotlib.org/stable/tutorials/colors/colormaps.html)

In [48]:
import matplotlib.pyplot as plt
import seaborn as sns

dates = df['date']

sns.set_context("talk")

_ = plt.xticks(rotation = 40) # This line rotates the ticks of the x-axis so that they are easy to read.
sns.histplot = (df,dates)

SyntaxError: invalid syntax (<ipython-input-48-74d74d6e12f0>, line 9)

&#9989; **Question 2.3.1 (2 points)**: Looking at the above output, answer the following questions: 

1. What country used nuclear weapons in 1960?
2. What country was the last to use nuclear weapons? Use some code to check that you had the right answer.


**<font size=+3>&#9998;</font> Do This**: *Record your answers here.*



---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (1 point)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "**Committing Part 2**", and push the changes to GitHub.

---

# Part 3: Object-Oriented Programming (38 points)



<img src=https://upload.wikimedia.org/wikipedia/commons/thumb/b/b5/Starting_position_in_a_chess_game.jpg/1024px-Starting_position_in_a_chess_game.jpg width=600px align=center >


Let's set up a fictional chess! Here is a [link](https://chess-teacher.com/wp-content/uploads/2016/01/symbols-of-chess-pieces.jpg) explaining the meaning of each piece.

&#9989; **Question 3.1 (9 points)**: The following cell contains a draft of a class called `ChessPiece` with only the `__init__` constructor. Complete the `__init__` constructor so that it:

- takes in three arguments, `color`, `shape`, with no default values. These arguments correspond to three attributes with the same name.

- creates another attribute called `position`. This should be a `tuple` set to `('a', 0)`. _Note:_ `'a'` is a `str` while `0` is an `int`.

- check whether `shape` is contained in the `list` `["bishop", "king", "knight", "pawn", "queen", "rook"]`. If no `raise` a `ValueError` with the message `"Piece does not exist."`.  [Hint](https://stackoverflow.com/questions/2052390/manually-raising-throwing-an-exception-in-python) 


In [56]:
# Put your code here
class ChessPiece:
    """Base class for chess pieces. 
    
    Attributes
    ----------
    color : str
        Color of the chess piece.
    
    shape : str
        Shape of the chess piece. It can be only one of the following
        ["bishop", "king", "knight", "pawn", "queen", "rook"]
    
    position : tuple
        Position on the chess board.
    
    """
    def __init__(self,color,shape, position = ('a',0)): 
        """Initialize the chess piece with its attributes color and shape 
        and set the initial position to the lower left corner of the board.
        
        Input Parameter
        ---------------
        color : str
            Color of the chess piece.
    
        shape : str
            Shape of the chess piece. It can be only one of the following
            ["bishop", "king", "knight", "pawn", "queen", "rook"]
        
        Raises
        ------
        ValueError
            For when the shape is not one of the possible shapes."""
        
        # Put your code here
        self.color = color
        self.shape = shape
        self.position = position
        pass 

    def val_error(shape):
        list = ["bishop", "king", "knight", "pawn", "queen", "rook"]
        if shape not in list:
            raise ValueError('Piece does not exsist'.format(shape=repr(shape)))
        


## Test your code

Let's that the code works. Run the following cells.

In [57]:
#
pawn = ChessPiece(color = "white", shape = "pawn")

pawn.position

('a', 0)

&#9989; **Question 3.1.1 (1 point):** Is `pawn` starting in the correct position `('a',0)`? If Yes move to the next, otherwise fix your code.

In [58]:
# Let's try breaking the code

archer = ChessPiece(color = "black", shape = "archer")

&#9989; **Question 3.1.2 (1 point):** Did you get a `ValueError` with the correct message? If Yes move to the next, otherwise fix your code.



---
&#9989; **Question 3.2 (21 points)**: In the code cell below there is a draft of a class called `Queen`. Complete this class which **should inherit** the `ChessPiece` class. In addition, modify the methods such that:

- the `__init__` constructor 
    * takes in one argument, `color`. 
    * set the `shape` attribute to `queen`
    * initializes the `position` attribute to `('d',1)` if the `color` attribute is `'white'` otherwise it initializes it to `('d',8)` if the color attribute is `'black'`.
    
- the `move_to` method should:
    * take in one argument `to_position`, with no default value. This will be the position in the chess board where the queen will move.
    * check whether the `to_position` is an _instance_ of a `tuple` [Hint](<https://letmegooglethat.com/?q=How+do+I+check+if+an+object+is+an+instance+of+a+given+class+python>).:
        * if no then it will raise a `TypeError` with the message `"Please pass a tuple."` [Hint](https://stackoverflow.com/questions/2052390/manually-raising-throwing-an-exception-in-python).
        
        * if yes then it will check if `to_position` is an allowed value, _i.e._ check if the elements of the `tuple` exist on the chess board. If the desired position does not exist on the board then raise a `ValueError` with the message `"Move not allowed."` Otherwise set the attribute `position` equal to `to_position`. _Note:_ the allowed x-coordinates are `"a"` through `"h"` and the allowed y-coordinates are `1` through `8`!.       
        

In [ ]:
# Put your code here
class Queen:
    """
    Queen chess piece. Can move in any direction. It inherits the ChessPiece class.
    """
    def __init__(): 
        """
        Initialize the chess piece with its attributes color and shape 
        and set the initial position to the lower left corner of the board.
        
        Input Parameter
        ---------------
        color : str
            Color of the chess piece.

        """
        # Put your code here
        pass
    
    def move_to():
        """
        Move the chess piece to the given position.
        
        Input parameter
        ---------------
        to_position: tuple
            Chess board coordinates where to move the piece.
        
        Raises
        ------
        ValueError
            For when the elements of the tuple are not part of the chess board.
        
        TypeError
            For when the input argument is not of the required type.   
        """
        # Put your code here
        pass

## Test your code

Let's check that the code works. Run the following cell.

In [ ]:
# Create two queens

harmon = Queen(color = 'black')
kasparov = Queen(color = "white")

print(harmon.position)
print(kasparov.position)


&#9989; **Question 3.3.1 (2 points):** Are the queens starting in the correct positions? Look at the board above. If Yes move to the next, otherwise fix your code.

In [ ]:
# Let's try to move 
harmon.move_to(("e",2))

print(harmon.position)

&#9989; **Question 3.3.2 (1 point)** Has the queen moved to the correct position? If Yes move to the next, otherwise fix your code.

In [ ]:
# Let's try and break the code

kasparov.move_to( "upper left by two")

&#9989; **Question 3.3.3 (1 point)**. Did you get a `TypeError` with the correct message? If Yes move to the next, otherwise fix your code.

In [ ]:
# Let's check the ValueError

harmon.move_to(to_position=("l",6))

&#9989; **Question 3.3.4 (1 point)**. Did you get a `ValueError` with the correct message? If Yes move to the next, otherwise fix your code.

---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (1 point)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "**Committing Part 3**", and push the changes to GitHub.

---

---
# Part 4. Conclusion (3 points)

Make sure all of your changes to your repository are committed and pushed to GitHub. 
Before you leave

1. Commit changes to your **local** `git` repository (1 point)

2. Push the changes to your GitHub repository (1 point)

3. Upload your notebook to D2L in case something went wrong with your repository or if you couldn't get the repository to work.  (1 point)

# You're done! Congrats on finishing your CMSE 202 Midterm!


![Moira2](https://media.giphy.com/media/d1E2HnwywoTkES08/giphy.gif)

In [1]:

# FINAL GRADE (Instructor's use only)
parts = [1, 9, 21, 38, 3]

total = sum(parts)
print(f"You final grade is {total}")

You final grade is 72


&#169; Copyright 2022,  Department of Computational Mathematics, Science and Engineering at Michigan State University